### ISSUE TITLE
Improve performance of EqClass

### ISSUE THREAD

**Issue by jdegoes** [scalaz/scalaz#1808]
Currently, `EqClass` is defined as so:

```scala
trait EqClass[A] {
  def equal(first: A, second: A): Boolean
}
```

Unfortunately, we have no guarantee that such implementations of `equal` will first check for reference equality before checking for value equality.

In order to _guarantee_ that reference equality is always checked first, we should refactor:

```scala
trait EqClass[A] {
  final def equal(l: A, r: A): Boolean = (l eq r) || valueEquals(l, r)

  protected def valueEquals(l: A, r: A): Boolean
}
```

Now we have a guarantee the high-performance shortcut will be used on all instances of `Eq`.

**Comment by jdegoes** [comment]
We should discuss laziness here, too, which also has an impact on performance.

**Comment by NeQuissimus** [comment]
You'd need to limit to `A <: AnyRef` to get `eq`.
What about `Eq` for primitives? Force boxing?

**Comment by jdegoes** [comment]
You can choose to not box:

```scala
@specialized 
trait EqClass[A] {
  final def equal(l: A, r: A): Boolean
}
```

But then you can't rely on `A` being an `AnyRef`, which means you can't use `eq`.

Hmmm. To not box or to use reference equality...

**Comment by jdegoes** [comment]
Perhaps numeric code can define its own specialized equals. Just thinking that for high-performance numerics, you can't really afford to box at the level of individual values...

For the average business application, though, reference equality shortcut seems like a far bigger win than unboxed primitive equality.

**Comment by hrhino** [comment]
Right. If I'm writing highly-optimized code, I'm probably going to be working only with concrete types, and `Eq` won't be a tool I reach for.

**Comment by jdegoes** [comment]
Also the APL folks have a point: doing everything at the level of arrays gives you lots of benefits. Could take a numerics lib in that direction...

**Comment by NeQuissimus** [comment]
Can the `eq` not be macro'd in somehow? If `A <: AnyRef`, add this blob...
(I know nothing about macros :D)

**Comment by NeQuissimus** [comment]
@jdegoes hmmm, I don't see how it could work with `instanceOf`... And the type alias `Eq[A]` cannot be `InstanceOf[EqualClass[A]]` because `InstanceOf` is invariant.

Not sure how one would solve that...

**Comment by jbgi** [comment]
I think it should work without problems.

But however I'd suggest a small variation:

```scala
package scalaz

sealed trait EqClass[T] {
  def equals(t1: T, t2: T): Boolean
}

trait EqAnyRef[T <: AnyRef] extends EqClass[T] {
  final def equals(t1: T, t2: T): Boolean = t1 eq t2 || valueEquals(t1, t2)
  protected def valueEquals(t1: T, t2: T): Boolean
}

@specialized
trait EqAnyVal[T <: AnyVal] extends EqClass[T]
```

**Comment by NeQuissimus** [comment]
I tried this on the weekend but I can't get this to work. I must be missing something.

- Since we'd want it to be a sealed trait, things such as `Ord` need to self-type on `EqClass` rather than extend
- In the `Prelude` I have no idea how to define `type Eq[A] = ???`. You can't do `type Eq[A] = InstanceOf[EqClass[A]]` because of `InstanceOf` being invariant.

**Comment by NeQuissimus** [comment]
This whole thread is a lot of ugly to take in... :D
Clearly, we need `sealed[scalaz]` :)

I was just thinking... We could also leave EqClass and only consider `AnyRef` a special case.
So people can extend `EqClass` if they want but `EqAnyRef` is the optimized case.
That way, there is no issue leaving `EqClass` unsealed, it actually has to be!

```scala
trait EqClass[A] {
  def equal(first: A, second: A): Boolean = first == second
}

trait EqAnyRef[A <: AnyRef] extends EqClass[A] {
  final override def equal(first: A, second: A): Boolean = (first eq second) || valueEqual(first, second)
  def valueEqual(first: A, second: A): Boolean
}

object Eq {
  implicit val stringEqual: EqClass[String] = new EqAnyRef[String] {
    def valueEqual(first: String, second: String) = first == second
  }

  @specialized
  object EqInt extends EqClass[Int] {}

  implicit val intEqual: EqClass[Int] = EqInt
}
```

**Comment by hamednourhani** [comment]
Hi @jdegoes can i work on this issue as first commit?

**Comment by NeQuissimus** [comment]
Go for it
